plugins/Login/Controller.php

```diff
@@ -307,10 +307,12 @@ protected function authenticateAndRedirect($login, $password, $urlToRedirect = f
             $redirectParams = UrlHelper::getArrayFromQueryString(UrlHelper::getQueryFromUrl($redirect));
             $module = Common::getRequestVar('module', '', 'string', $redirectParams);
             // when module is login, we redirect to home...
-            if ($module !== 'Login' && $module !== Piwik::getLoginPluginName() && $redirect) {
+            if (!empty($module) && $module !== 'Login' && $module !== Piwik::getLoginPluginName() && $redirect) {
                 $host = Url::getHostFromUrl($redirect);
                 // we only redirect to a trusted host
-                if ($host && Url::isValidHost($host)) {
+                if ($host == Url::getHost()
```

Comment:
we probably still need to check `!empty($host)` as `getHost()` may return an empty host when the config setting is disabled for example. It would be able to trick matomo into redirecting maybe when providing some weird URL although I don't really think it's possible and it should be all fine without the !empty but better be safe maybe.